### PR TITLE
python3Packages.{irctokens,ircstates,ircrobots}: init

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

I'm writing software depending on these libraries.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Result of `nixpkgs-review pr 99521` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python37Packages.anyio</li>
    <li>python37Packages.async_stagger</li>
    <li>python37Packages.asyncio-throttle</li>
    <li>python37Packages.ircrobots</li>
    <li>python37Packages.ircstates</li>
    <li>python37Packages.irctokens</li>
    <li>python38Packages.anyio</li>
    <li>python38Packages.async_stagger</li>
    <li>python38Packages.asyncio-throttle</li>
    <li>python38Packages.ircrobots</li>
    <li>python38Packages.ircstates</li>
    <li>python38Packages.irctokens</li>
  </ul>
</details>

File: pkgs/development/python-modules/async_stagger/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    sha256 = "007l54fbk2dfzv3vmqz98m1i37mzxkkva5r4fiwq2pg8nb61fy0w";
+  };
+
+  checkInputs = [
+    pytestCheckHook
+    pytest-asyncio
+    pytest-mock
+  ];
+
+  disabledTests = lib.optional (!pythonOlder "3.8") [
```
Comment (jonringer) [inline pkgs/development/python-modules/async_stagger/default.nix:26]: this is of shape `[[string]]`
```suggestion
  disabledTests = lib.optionals (!pythonOlder "3.8") [
```

File: pkgs/development/python-modules/asyncio-throttle/default.nix [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  };
+
+  checkInputs = [
+    pytest-asyncio
+    #pytest-trio
+    #pytest-tornasync
+    #pytest-twisted
+    pytestCheckHook
+    #twisted
+  ];
```
Comment (jonringer) [inline pkgs/development/python-modules/asyncio-throttle/default.nix:28]: remove if not needed
```suggestion
  checkInputs = [
    pytest-asyncio
    pytestCheckHook
  ];
```

File: pkgs/development/python-modules/ircrobots/default.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    ./relax-dependencies.patch
+  ];
+
+  propagatedBuildInputs = [
+    anyio
+    asyncio-throttle
+    ircstates
+    async_stagger
+    async-timeout
+  ] ++ lib.optional (pythonOlder "3.7") [
```
Comment (jonringer) [inline pkgs/development/python-modules/ircrobots/default.nix:36]:
```suggestion
  ] ++ lib.optionals (pythonOlder "3.7") [
```

File: pkgs/development/python-modules/asyncio-throttle/default.nix (hunk 2 above)
Comment (mweinelt) [reply pkgs/development/python-modules/asyncio-throttle/default.nix:28]: Leftovers, had to choose one runtime to run the tests with. Removed!

File: pkgs/development/python-modules/async_stagger/default.nix (hunk 1 above)
Comment (mweinelt) [reply pkgs/development/python-modules/async_stagger/default.nix:26]: I always get those wrong. How do you remember which is which?

Comment (jonringer) [reply pkgs/development/python-modules/async_stagger/default.nix]: optionals -> "many" -> an array
optional   -> "one" -> singleton

File: pkgs/development/python-modules/anyio/default.nix [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    owner = "agronholm";
+    repo = pname;
+    rev = version;
+    sha256 = "06nazfrm2sclp3lpgsn9wl8vmqxvx36s3gr2gnqz3zhjpf3glkxv";
+  };
+
+  propagatedBuildInputs = [
+    idna
+    sniffio
+  ] ++ lib.optional (pythonOlder "3.8") [
```
Comment (jonringer) [inline pkgs/development/python-modules/anyio/default.nix:32]: missed one ;)
```suggestion
  ] ++ lib.optionals (pythonOlder "3.8") [
```

Comment (mweinelt) [reply pkgs/development/python-modules/anyio/default.nix]: Oh, fixed btw.

**Comment by mweinelt** [comment]
Result of `nixpkgs-review pr 99521` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python37Packages.anyio</li>
    <li>python37Packages.async_stagger</li>
    <li>python37Packages.asyncio-throttle</li>
    <li>python37Packages.ircrobots</li>
    <li>python37Packages.ircstates</li>
    <li>python37Packages.irctokens</li>
    <li>python38Packages.anyio</li>
    <li>python38Packages.async_stagger</li>
    <li>python38Packages.asyncio-throttle</li>
    <li>python38Packages.ircrobots</li>
    <li>python38Packages.ircstates</li>
    <li>python38Packages.irctokens</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 99521` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python37Packages.anyio</li>
    <li>python37Packages.ircrobots</li>
    <li>python38Packages.anyio</li>
    <li>python38Packages.ircrobots</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>python37Packages.async_stagger</li>
    <li>python37Packages.asyncio-throttle</li>
    <li>python37Packages.ircstates</li>
    <li>python37Packages.irctokens</li>
    <li>python38Packages.async_stagger</li>
    <li>python38Packages.asyncio-throttle</li>
    <li>python38Packages.ircstates</li>
    <li>python38Packages.irctokens</li>
  </ul>
</details>

**Comment by mweinelt** [comment]
@SuperSandro2000 Thanks. Your comment needs more build logs to be actionable.

**Comment by SuperSandro2000** [comment]
Sandbox failure:

```
Sourcing python-recompile-bytecode-hook.sh
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pip-build-hook
Using pipBuildPhase
Using pipShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing pytest-check-hook
Using pytestCheckPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/3k5dsgfygla44zx1ic2faqbw497x2x33-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/tox.ini
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
Executing pipBuildPhase
Creating a wheel...
[33mWARNING: The directory '/homeless-shelter/Library/Caches/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.[0m
Processing /private/tmp/nix-build-python3.8-anyio-2.0.2.drv-0/source
    Preparing wheel metadata ... [?25l- done
[?25hBuilding wheels for collected packages: anyio
  Building wheel for anyio (PEP 517) ... [?25l- \ done
[?25h  Created wheel for anyio: filename=anyio-0.0.0-py3-none-any.whl size=62720 sha256=1b4c41896c61e82f6c909d46890da4cda7c60dcfe9bad18d2df81b4129369c6c
  Stored in directory: /private/tmp/nix-build-python3.8-anyio-2.0.2.drv-0/pip-ephem-wheel-cache-kssi843v/wheels/c4/13/0b/a4779543c6f33fea8f7793da32f41c8acb15906ae23cb3e287
Successfully built anyio
Finished creating a wheel...
Finished executing pipBuildPhase
@nix { "action": "setPhase", "phase": "installPhase" }
installing
Executing pipInstallPhase
/private/tmp/nix-build-python3.8-anyio-2.0.2.drv-0/source/dist /private/tmp/nix-build-python3.8-anyio-2.0.2.drv-0/source
Processing ./anyio-0.0.0-py3-none-any.whl
Requirement already satisfied: idna>=2.8 in /nix/store/gh2zhfglh81q72hb3d0pgypp5ri9ry30-python3.8-idna-2.10/lib/python3.8/site-packages (from anyio==0.0.0) (2.10)
Requirement already satisfied: sniffio>=1.1 in /nix/store/544kydnyjgvhc1zpfxwglw03r10zj08g-python3.8-sniffio-1.1.0/lib/python3.8/site-packages (from anyio==0.0.0) (1.1.0)
Installing collected packages: anyio
Successfully installed anyio-0.0.0
/private/tmp/nix-build-python3.8-anyio-2.0.2.drv-0/source
Finished executing pipInstallPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
post-installation fixup
strip is /nix/store/1h9ivag16gyysxjgfwq5dfk8w6g246g7-cctools-binutils-darwin-949.0.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/wdgvpvlkxlp5ajm7a4qamcjmx7kkbnnz-python3.8-anyio-2.0.2/lib 
patching script interpreter paths in /nix/store/wdgvpvlkxlp5ajm7a4qamcjmx7kkbnnz-python3.8-anyio-2.0.2
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
@nix { "action": "setPhase", "phase": "installCheckPhase" }
running install tests
no Makefile or custom buildPhase, doing nothing
@nix { "action": "setPhase", "phase": "pythonCatchConflictsPhase" }
pythonCatchConflictsPhase
@nix { "action": "setPhase", "phase": "pythonRemoveBinBytecodePhase" }
pythonRemoveBinBytecodePhase
@nix { "action": "setPhase", "phase": "pythonImportsCheckPhase" }
pythonImportsCheckPhase
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: anyio
@nix { "action": "setPhase", "phase": "pytestCheckPhase" }
pytestCheckPhase
Executing pytestCheckPhase
[1m============================= test session starts ==============================[0m
platform darwin -- Python 3.8.6, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: /private/tmp/nix-build-python3.8-anyio-2.0.2.drv-0/source, inifile: setup.cfg, testpaths: tests
plugins: anyio-0.0.0, curio-1.2, hypothesis-5.30.0
[1mcollecting ... [0m[1m
collected 518 items                                                            [0m                                       [  5%][0m
tests/test_signals.py                                              [  6%][0m
tests/test_subprocesses.py 
=================================== FAILURES ===================================
[31m[1m___________________ TestTLSStream.test_send_receive[asyncio] ___________________[0m
[1m[31mtests/streams/test_tls.py[0m:27: in test_send_receive
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_______________ TestTLSStream.test_send_receive[asyncio+uvloop] ________________[0m
[1m[31mtests/streams/test_tls.py[0m:27: in test_send_receive
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m____________________ TestTLSStream.test_send_receive[curio] ____________________[0m
[1m[31mtests/streams/test_tls.py[0m:27: in test_send_receive
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m____________________ TestTLSStream.test_send_receive[trio] _____________________[0m
[1m[31mtests/streams/test_tls.py[0m:27: in test_send_receive
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_________________ TestTLSStream.test_extra_attributes[asyncio] _________________[0m
[1m[31mtests/streams/test_tls.py[0m:55: in test_extra_attributes
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_____________ TestTLSStream.test_extra_attributes[asyncio+uvloop] ______________[0m
[1m[31mtests/streams/test_tls.py[0m:55: in test_extra_attributes
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m__________________ TestTLSStream.test_extra_attributes[curio] __________________[0m
[1m[31mtests/streams/test_tls.py[0m:55: in test_extra_attributes
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m__________________ TestTLSStream.test_extra_attributes[trio] ___________________[0m
[1m[31mtests/streams/test_tls.py[0m:55: in test_extra_attributes
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m______________________ TestTLSStream.test_unwrap[asyncio] ______________________[0m
[1m[31mtests/streams/test_tls.py[0m:93: in test_unwrap
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m__________________ TestTLSStream.test_unwrap[asyncio+uvloop] ___________________[0m
[1m[31mtests/streams/test_tls.py[0m:93: in test_unwrap
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_______________________ TestTLSStream.test_unwrap[curio] _______________________[0m
[1m[31mtests/streams/test_tls.py[0m:93: in test_unwrap
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_______________________ TestTLSStream.test_unwrap[trio] ________________________[0m
[1m[31mtests/streams/test_tls.py[0m:93: in test_unwrap
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_________________ TestTLSStream.test_alpn_negotiation[asyncio] _________________[0m
[1m[31mtests/streams/test_tls.py[0m:125: in test_alpn_negotiation
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_____________ TestTLSStream.test_alpn_negotiation[asyncio+uvloop] ______________[0m
[1m[31mtests/streams/test_tls.py[0m:125: in test_alpn_negotiation
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m__________________ TestTLSStream.test_alpn_negotiation[curio] __________________[0m
[1m[31mtests/streams/test_tls.py[0m:125: in test_alpn_negotiation
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m__________________ TestTLSStream.test_alpn_negotiation[trio] ___________________[0m
[1m[31mtests/streams/test_tls.py[0m:125: in test_alpn_negotiation
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m____________ TestTLSStream.test_ragged_eofs[asyncio-both_standard] _____________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m___________ TestTLSStream.test_ragged_eofs[asyncio-server_standard] ____________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m___________ TestTLSStream.test_ragged_eofs[asyncio-client_standard] ____________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m___________ TestTLSStream.test_ragged_eofs[asyncio-neither_standard] ___________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_________ TestTLSStream.test_ragged_eofs[asyncio+uvloop-both_standard] _________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m________ TestTLSStream.test_ragged_eofs[asyncio+uvloop-server_standard] ________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m________ TestTLSStream.test_ragged_eofs[asyncio+uvloop-client_standard] ________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_______ TestTLSStream.test_ragged_eofs[asyncio+uvloop-neither_standard] ________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_____________ TestTLSStream.test_ragged_eofs[curio-both_standard] ______________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m____________ TestTLSStream.test_ragged_eofs[curio-server_standard] _____________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m____________ TestTLSStream.test_ragged_eofs[curio-client_standard] _____________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m____________ TestTLSStream.test_ragged_eofs[curio-neither_standard] ____________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m______________ TestTLSStream.test_ragged_eofs[trio-both_standard] ______________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_____________ TestTLSStream.test_ragged_eofs[trio-server_standard] _____________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_____________ TestTLSStream.test_ragged_eofs[trio-client_standard] _____________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m____________ TestTLSStream.test_ragged_eofs[trio-neither_standard] _____________[0m
[1m[31mtests/streams/test_tls.py[0m:169: in test_ragged_eofs
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m______________ TestTLSStream.test_receive_send_after_eof[asyncio] ______________[0m
[1m[31mtests/streams/test_tls.py[0m:199: in test_receive_send_after_eof
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m__________ TestTLSStream.test_receive_send_after_eof[asyncio+uvloop] ___________[0m
[1m[31mtests/streams/test_tls.py[0m:199: in test_receive_send_after_eof
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_______________ TestTLSStream.test_receive_send_after_eof[curio] _______________[0m
[1m[31mtests/streams/test_tls.py[0m:199: in test_receive_send_after_eof
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[31m[1m_______________ TestTLSStream.test_receive_send_after_eof[trio] ________________[0m
[1m[31mtests/streams/test_tls.py[0m:199: in test_receive_send_after_eof
    server_sock.bind(('127.0.0.1', 0))
[1m[31mE   PermissionError: [Errno 1] Operation not permitted[0m
[33m=============================== warnings summary ===============================[0m
/nix/store/bgc2jqndjpn6g5i4njrww4n28vrv2shc-python3.8-curio-1.2/lib/python3.8/site-packages/curio/pytest_plugin.py:60: 115 tests with warnings
  /nix/store/bgc2jqndjpn6g5i4njrww4n28vrv2shc-python3.8-curio-1.2/lib/python3.8/site-packages/curio/pytest_plugin.py:60: PytestDeprecationWarning: direct construction of Function has been deprecated, please use Function.from_parent
    item = pytest.Function(name, parent=collector)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
[31m================ [31m[1m36 failed[0m, [32m482 passed[0m, [33m115 warnings[0m[31m in 22.24s[0m[31m =================[0m
```

**Pull request merged**